### PR TITLE
extract RunCacheProbe function

### PR DESCRIPTION
**Pull request by johnstcn**
Part of https://github.com/coder/team-coconut/issues/11

Builds on top of https://github.com/coder/envbuilder/pull/282 and https://github.com/coder/envbuilder/pull/283

Extracts the logic for `--get-cached-image` to a separate function

Also pulls out some other common logic shared between `Run` and `RunCacheProbe`. There is more that could be DRYed up, but leaving this for later.

### CODE REVIEW

File: envbuilder.go [hunk 1]
```diff
@@ -1137,3 +1407,111 @@ func copyFile(src, dst string) error {
 	}
 	return nil
 }
+
```
Comment (johnstcn) [inline envbuilder.go:1410]: review: below functions are extracted

File: envbuilder.go [hunk 2]
```diff
@@ -902,6 +838,340 @@ ENTRYPOINT [%q]`, exePath, exePath, exePath)
 	return nil
 }
 
+// RunCacheProbe performs a 'dry-run' build of the image and checks that
+// all of the resulting layers are present in options.CacheRepo.
+// Logger is the logf to use for all operations.
+// Filesystem is the filesystem to use for all operations.
+// Defaults to the host filesystem.
```
Comment (dannykopping) [inline envbuilder.go:845]: Were these arguments previous, but you later moved them to options.Options by any chance?
This comment feels potentially stale.

File: envbuilder.go [hunk 3]
```diff
@@ -902,6 +838,340 @@ ENTRYPOINT [%q]`, exePath, exePath, exePath)
...
+func RunCacheProbe(ctx context.Context, opts options.Options) (v1.Image, error) {
+	if !opts.GetCachedImage {
+		return nil, fmt.Errorf("developer error: RunCacheProbe must be run with --get-cached-image")
+	}
+	if opts.CacheRepo == "" {
+		return nil, fmt.Errorf("--cache-repo must be set when using --get-cached-image")
```
Comment (dannykopping) [inline envbuilder.go:851]: Nit: inconsistent use of `xerrors.Errorf` and `fmt.Errorf`; was this intentional?

File: envbuilder.go [hunk 4]
```diff
@@ -902,6 +838,340 @@ ENTRYPOINT [%q]`, exePath, exePath, exePath)
...
+	}
+
+	stageNumber := 0
+	startStage := func(format string, args ...any) func(format string, args ...any) {
+		now := time.Now()
+		stageNumber++
```
Comment (dannykopping) [inline envbuilder.go:857]: Nit: pass in an int pointer instead perhaps?

File: envbuilder.go [hunk 5]
```diff
@@ -902,6 +838,340 @@ ENTRYPOINT [%q]`, exePath, exePath, exePath)
...
+	caBundle, err := initCABundle(opts.SSLCertBase64)
+	if err != nil {
+		return nil, err
+	}
+
+	cleanupDockerConfigJSON, err := initDockerConfigJSON(opts.DockerConfigBase64)
+	if err != nil {
+		return nil, err
+	}
+	defer func() { _ = cleanupDockerConfigJSON() }() // best effort
```
Comment (dannykopping) [inline envbuilder.go:877]: Perhaps log the error at least?

File: envbuilder.go [hunk 6]
```diff
@@ -902,6 +838,340 @@ ENTRYPOINT [%q]`, exePath, exePath, exePath)
...
+			newColor(color.FgCyan).Sprintf(opts.WorkspaceFolder),
+		)
+
+		reader, writer := io.Pipe()
+		defer reader.Close()
+		defer writer.Close()
+		go func() {
+			data := make([]byte, 4096)
+			for {
+				read, err := reader.Read(data)
```
Comment (dannykopping) [inline envbuilder.go:893]: Nit: consider using https://pkg.go.dev/bufio#NewReaderSize with `ReadBytes(delim byte)` where `delim` is '\r'?

File: envbuilder.go [hunk 7]
```diff
@@ -902,6 +838,340 @@ ENTRYPOINT [%q]`, exePath, exePath, exePath)
...
+
+	stdoutReader, stdoutWriter := io.Pipe()
+	stderrReader, stderrWriter := io.Pipe()
+	defer stdoutReader.Close()
+	defer stdoutWriter.Close()
+	defer stderrReader.Close()
+	defer stderrWriter.Close()
+	go func() {
+		scanner := bufio.NewScanner(stdoutReader)
+		for scanner.Scan() {
```
Comment (dannykopping) [inline envbuilder.go:1089]: Should these goroutines exit when the context is done?

File: envbuilder.go [hunk 8]
```diff
@@ -902,6 +838,340 @@ ENTRYPOINT [%q]`, exePath, exePath, exePath)
...
+		CacheRunLayers:    true,
+		CacheCopyLayers:   true,
+		CompressedCaching: true,
+		Compression:       config.ZStd,
+		// Maps to "default" level, ~100-300 MB/sec according to
+		// benchmarks in klauspost/compress README
+		// https://github.com/klauspost/compress/blob/67a538e2b4df11f8ec7139388838a13bce84b5d5/zstd/encoder_options.go#L188
+		CompressionLevel: 3,
+		CacheOptions: config.CacheOptions{
+			// Cache for a week by default!
```
Comment (dannykopping) [inline envbuilder.go:1131]: Nit: this comment is out of step with the definition, and could rot

**Review by dannykopping — APPROVED** [review]
LGTM, mostly small nits and suggestions
Quite a lot of complexity in the `RunCacheProbe` function which would be good to decompose later on.

File: envbuilder.go (hunk 2 above)
Comment (johnstcn) [reply envbuilder.go:845]: Yep, this is stale 👍

File: envbuilder.go (hunk 3 above)
Comment (johnstcn) [reply envbuilder.go:851]: Most likely; fmt.Errorf is the majority here so will standardize on that.

File: envbuilder.go (hunk 4 above)
Comment (johnstcn) [reply envbuilder.go:857]: This is a pre-existing non-pointer integer; I want to refactor this whole stage builder business later on. A little copying for now.

File: envbuilder.go (hunk 6 above)
Comment (johnstcn) [reply envbuilder.go:893]: pre-existing; will address in follow-up

File: envbuilder.go (hunk 7 above)
Comment (johnstcn) [reply envbuilder.go:1089]: good point; we didn't need to worry about this before because the binary would just exit

File: envbuilder.go (hunk 6 above)
Comment (mtojek) [reply envbuilder.go:874]: +1

**Review by mtojek — COMMENTED** [review]
Please ignore comments that refer to the original code.

File: envbuilder.go [hunk 9]
```diff
@@ -902,6 +822,336 @@ ENTRYPOINT [%q]`, exePath, exePath, exePath)
...
+			AllowedPaths:    nil,
+		})
+	}
+
+	stdoutReader, stdoutWriter := io.Pipe()
+	stderrReader, stderrWriter := io.Pipe()
+	defer stdoutReader.Close()
+	defer stdoutWriter.Close()
+	defer stderrReader.Close()
+	defer stderrWriter.Close()
```
[the remark this answers is not in the document]
Comment (johnstcn) [reply envbuilder.go:1067]: pre-existing, but probably yes

File: envbuilder.go [hunk 10]
```diff
@@ -902,6 +822,336 @@ ENTRYPOINT [%q]`, exePath, exePath, exePath)
...
+// all of the resulting layers are present in options.CacheRepo.
+func RunCacheProbe(ctx context.Context, opts options.Options) (v1.Image, error) {
+	if !opts.GetCachedImage {
+		return nil, fmt.Errorf("developer error: RunCacheProbe must be run with --get-cached-image")
+	}
+	if opts.CacheRepo == "" {
+		return nil, fmt.Errorf("--cache-repo must be set when using --get-cached-image")
+	}
+
+	stageNumber := 0
```
[the remark this answers is not in the document]
Comment (johnstcn) [reply envbuilder.go:840]: will rename in a follow-up

File: envbuilder.go [hunk 11]
```diff
@@ -902,6 +822,336 @@ ENTRYPOINT [%q]`, exePath, exePath, exePath)
...
+		ForceUnpack:       true,
+		BuildArgs:         buildParams.BuildArgs,
+		CacheRepo:         opts.CacheRepo,
+		Cache:             opts.CacheRepo != "" || opts.BaseImageCacheDir != "",
+		DockerfilePath:    buildParams.DockerfilePath,
+		DockerfileContent: buildParams.DockerfileContent,
+		RegistryOptions: config.RegistryOptions{
+			Insecure:      opts.Insecure,
+			InsecurePull:  opts.Insecure,
+			SkipTLSVerify: opts.Insecure,
```
[the remark this answers is not in the document]
Comment (johnstcn) [reply envbuilder.go:1123]: What is the issue here?

File: envbuilder.go [hunk 12]
```diff
@@ -902,6 +822,336 @@ ENTRYPOINT [%q]`, exePath, exePath, exePath)
...
+
+	endStage := startStage("🏗️ Checking for cached image...")
+	image, err := executor.DoCacheProbe(kOpts)
+	if err != nil {
+		return nil, fmt.Errorf("get cached image: %w", err)
+	}
+	endStage("🏗️ Found cached image!")
+
+	// Sanitize the environment of any opts!
+	options.UnsetEnv()
```
[the remark this answers is not in the document]
Comment (johnstcn) [reply envbuilder.go:1144]: added at beginning of func

File: envbuilder.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (mtojek) [reply envbuilder.go:1125]: ignoring TLS issues

**Review by mtojek — APPROVED** [review]

File: envbuilder.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (johnstcn) [reply envbuilder.go:1125]: I think it's intentional here.

File: envbuilder.go [hunk 13]
```diff
@@ -1123,17 +1373,125 @@ func maybeDeleteFilesystem(logger log.Func, force bool) error {
...
+				"rootdirectory": layerCacheDir,
+			},
+		},
+	}
+	cfg.Log.Level = "error"
+
+	// Spawn an in-memory registry to cache built layers...
+	registry := handlers.NewApp(ctx, cfg)
+
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
```
[the remark this answers is not in the document]
Comment (johnstcn) [reply envbuilder.go:1469]: I don't recall ever needing to do this; you can always inspect the local dir contents

File: envbuilder.go [hunk 14]
```diff
@@ -1123,17 +1373,125 @@ func maybeDeleteFilesystem(logger log.Func, force bool) error {
...
+	if dockerConfigBase64 == "" {
+		return noop, nil
+	}
+	cfgPath := filepath.Join(constants.MagicDir, "config.json")
+	decoded, err := base64.StdEncoding.DecodeString(dockerConfigBase64)
+	if err != nil {
+		return noop, fmt.Errorf("decode docker config: %w", err)
+	}
+	var configFile DockerConfig
+	decoded, err = hujson.Standardize(decoded)
```
[the remark this answers is not in the document]
Comment (johnstcn) [reply envbuilder.go:1423]: :D

File: envbuilder.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (mtojek) [reply envbuilder.go:1125]: synced offline: I assumed incorrectly that it is `SkipTLSVerify = true`

File: envbuilder.go [hunk 15]
```diff
@@ -1123,17 +1373,125 @@ func maybeDeleteFilesystem(logger log.Func, force bool) error {
...
+	ok := certPool.AppendCertsFromPEM(data)
+	if !ok {
+		return nil, fmt.Errorf("failed to append the ssl cert to the global pool: %s", data)
+	}
+	return data, nil
+}
+
+func initDockerConfigJSON(dockerConfigBase64 string) (func() error, error) {
+	var cleanupOnce sync.Once
+	noop := func() error { return nil }
```
[the remark this answers is not in the document]
Comment (johnstcn) [reply envbuilder.go:1413]: Returning a `noop` function removes the need for the caller to care whether it's nil or not. I much prefer this pattern personally.

File: envbuilder.go (hunk 1 above)
Comment (johnstcn) [reply envbuilder.go:1391]: Yeah we probably should; I'd like to simplify the `envbuilder.Run` / `RunCacheProbe` functions as much as possible for future maintenance.

**Pull request merged**